lib/OpenQA/WebAPI/Controller/Test.pm

```diff
@@ -421,6 +440,28 @@ sub _show {
     $self->render('test/result');
 }
 
+sub badge ($self) { $self->_badge($self->_get_current_job(1)) }
+
+sub _badge ($self, $job) {
+    my $badge_text = 'Error 404: Job not found!';
+    my $badge_color = $BADGE_RESULT_COLORS{cancelled};
+    my $status = 404;
+
+    if ($job) {
+        my $result = ($job->result eq 'none') ? $job->state : $job->result;
+        if ($result eq 'scheduled' && $job->blocked_by_id) {
+            $result = 'blocked';
```

Comment:
Those three lines look like the apply a result based on different factors but in one line you use the ternary operator and then the `if`. I would suggest to stay consistent with one approach. However I think the problem is actually that you are mixing states and results inconsistently with the definition from lib/OpenQA/Jobs/Constants.pm . By the way, another point is that you should use results from there and not the verbatim strings please.

@Martchus what is your take on the mixture for states and results?

Reply:
> @Martchus what is your take on the mixture for states and results?

I guess for the sake of these "badges" I guess it is ok. If there's a result then the user likely cares about the result (and not the state which would only be the distinction between cancelled/done). If there's no result the user cares about the state. It is also fine to take the "blocked"-flag into account.

Maybe these three lines could be extracted into a separate function, though. I'd call it `sub concise_result`. By the way, I'd also change the coding style a little bit, use constants and avoid repeated calls to the same functions (because function calls are slow in Perl):
```
sub concise_result ($self) {
    my $res = $job->result;
    $res = $job->state if $res eq NONE;
    $res = BLOCKED if $res eq SCHEDULED && $job->blocked_by_id;
}
```

Reply:
The mix of state and result comes from the openQA code actually.
You can find it at several places, eg. assets/javascripts/test_result.js line 1007.

Reply:
I know, we already have something similar in JavaScript. It would still be nice to extract it into its own function. Maybe even add a comment that this is in line with some other JavaScript function (and vice versa).

Reply:
We could move this `concise_result` method to the job schema and call it from there.
Then we could remove the extra js code as well.

Reply:
I haven't looked at the JavaScript code yet. Maybe this could be computed server-side as well, maybe it is better to leave it client-side (e.g. when it is computed within a DataTables rendering hook).